staging/src/k8s.io/sample-controller/pkg/apis/example/v1alpha1/types.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=nginx
+
+// NGINX is a specification for a NGINX resource
+type NGINX struct {
```

Comment:
Don't we want internal types as well? :thinking:

Reply:
And hence generating the `zz_generated.conversion.go` files as well.

Reply:
So if users are creating a CRD controller, conversions and internal types should not be required. Internal types & the conversions between versions are only useful when creating a custom API server (and are only used by the custom apiserver, iiuc)

The internal API version & conversions would never actually be used in the controller, so I thought it best to omit them to show users what is/isn't required. Maybe we should include a note in `pkg/apis/example/v1alpha1/types.go` explaining why we don't use the other generators?

Reply:
>  The internal API version & conversions would never actually be used in the controller, so I thought it best to omit them to show users what is/isn't required. Maybe we should include a note in pkg/apis/example/v1alpha1/types.go explaining why we don't use the other generators?

Makes sense. :+1: